src/core/types.rs

```diff
@@ -0,0 +1,303 @@
+            if node.depth.is_none() {
+                node.depth = Some(current_depth);
+
+                // Make sure we process the dependencies of this node
+                to_process.extend(node.dependencies.iter().cloned());
+                current_depth += 1;
+            }
+        }
+
+        // Finally merge graphs and add dependency on proper nodes
+        self.merge(graph);
+        self.add_dependency(from, on); // On will now be a part of the graph
+    }
+
+    // TODO: Read up on the `where` syntax
+    pub fn with<F>(self, name: &str, f: F) where F: Fn(&DepNode) {
+        let node = self.nodes.get(name).unwrap().borrow();
+        // TODO: Not sure I understand the deref part
+        f(node.deref())
+    }
+
+    pub fn has_node(&self, name: &str) -> bool {
+        self.nodes.contains_key(name)
+    }
+
+    pub fn clone_from(&self, name: &str) -> DepGraph {
+        assert!(self.has_node(name));
+        let mut clone = DepGraph::new();
+
+        let mut to_clone = BTreeSet::new();
+        to_clone.insert(name.to_string());
+        while let Some(name) = to_clone.pop_first() {
+            let node = self.nodes.get(&name).unwrap().borrow();
+            clone.add(node.clone());
+
+            // Mark all dependencies which are a not yet in the cloned graph as needed to clone
+            for dep in &node.dependencies {
+                if !clone.has_node(&dep) {
+                    to_clone.insert(dep.clone());
+                }
+            }
+        }
+
+        // We now have a subset of nodes, remove any node `dependents` which were not copied over
+        for (_, node_cell) in &clone.nodes {
+            let mut node = node_cell.borrow_mut();
+
+            // Remove non existent dependents
+            // TODO: Docs show |&dep| like signature, but that failed?
+            node.dependents.retain(|dep| clone.has_node(dep.as_str()));
+        }
+
+        return clone
+    }
+
+    pub fn merge(&mut self, other: DepGraph) {
+        for (name, node_cell) in other.nodes {
+            if self.nodes.contains_key(&name) {
+                // Just update the data
+                let mut existing = self.nodes.get(&name).unwrap().borrow_mut();
+                existing.merge(node_cell.into_inner());
+            } else {
+                // Add the whole node
+                self.add(node_cell.into_inner());
+            }
+        }
+    }
+}
+
+
+#[pymethods]
+impl DepGraph {
+    pub fn size(&self) -> usize {
+        self.nodes.len()
+    }
+
+    pub fn num_dependencies(&self) -> usize {
+        let mut acc = 0;
+        for (_, node_cell) in &self.nodes {
+            let node = node_cell.borrow();
+            acc += node_cell.borrow().dependencies.len();
+        }
+
+        return acc
+    }
+
+    pub fn keys(&self) -> HashSet<String> {
+        // TODO: Probably expensive, faster to write a conversion for graph.keys() and Keys type?
+        self.nodes.iter().map(|(key, _)| key.to_string()).collect()
```

Comment:
I believe you can do:
```
self.nodes.keys().cloned().collect()
```

May be a bit faster since its a built-in implementation of this (sidenote - clippy would suggest to do this)